// @mux/cli/lib/commands/sign.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const chalk = require("chalk");
const clipboard = require("clipboardy");
const base_1 = require("../command-bases/base");
class Sign extends base_1.default {
    async run() {
        const { args, flags } = this.parse(Sign);
        const playbackId = args['playback-id'];
        const options = { expiration: flags.expiresIn, type: flags.type };
        const key = this.JWT.sign(playbackId, options);
        const url = `https://stream.mux.com/${playbackId}.m3u8?token=${key}`;
        this.log(chalk `
🔑 {bold.underline Your Mux Signed Token}
{blue ${key}}

🌏 {bold.underline Full URL}
{green ${url}}
`);
        try {
            await clipboard.write(url);
            this.log(`👉 Copied Full URL to your system clipboard`);
        }
        catch (_a) {
            this.error('Unable to copy full url automatically');
        }
    }
}
exports.default = Sign;
Sign.description = 'Creates a new signed URL token for a playback ID';
Sign.args = [
    {
        name: 'playback-id',
        description: 'Playback ID to create a signed URL token for.',
        required: true,
    },
];
Sign.flags = {
    expiresIn: command_1.flags.string({
        char: 'e',
        description: 'How long the signature is valid for. If no unit is specified, milliseconds is assumed.',
        default: '7d',
    }),
    type: command_1.flags.string({
        char: 't',
        description: 'What type of token this signature is for.',
        default: 'video',
        options: ['video', 'thumbnail', 'gif', 'storyboard'],
    })
};
