// @mux/cli/lib/commands/live/complete.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Listr = require("listr");
const command_1 = require("@oclif/command");
const live_base_1 = require("../../command-bases/live-base");
const chalk = require("chalk");
class LiveComplete extends live_base_1.default {
    async run() {
        const { args, flags } = this.parse(LiveComplete);
        const streamId = this.getStreamId(flags, args.streamName);
        try {
            await (new Listr([
                {
                    title: `Signaling completion of '${streamId}'`,
                    task: async (ctx, task) => {
                        await this.Video.LiveStreams.signalComplete(streamId),
                            task.title = `${task.title} (OK)`;
                    },
                },
                {
                    title: `Disabling '${streamId}'`,
                    enabled: () => flags.disableAfterCompletion,
                    task: async (ctx, task) => {
                        await this.Video.LiveStreams.disable(streamId);
                        task.title = `${task.title} (OK)`;
                    },
                },
            ], {}).run());
        }
        catch (err) {
            // TODO: make this clearer
            console.log(chalk.redBright('Error:') +
                "\n\n" +
                err);
            if (err instanceof Error) {
                this.error(err);
            }
            else {
                throw err;
            }
        }
    }
}
exports.default = LiveComplete;
LiveComplete.description = "Signal to Mux that a live stream has concluded and should be closed.";
LiveComplete.args = [
    ...live_base_1.default.argsForSingleLiveStream,
];
LiveComplete.flags = Object.assign(Object.assign({}, live_base_1.default.flagsForSingleLiveStream), { disableAfterCompletion: command_1.flags.boolean({
        name: 'disable-after-completion',
        char: 'd',
        description: 'If set, disables the stream upon completion.',
        default: false,
    }) });
